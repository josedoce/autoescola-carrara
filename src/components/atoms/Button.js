import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const ButtonColors = {
  default: 'default',
  primary: 'primary',
  danger: 'danger',
};

export const ButtonsVariants = {
  default: 'default',
  outlined: 'outlined',
  link: 'link'
}

const getMainColor = ({theme, color})=> {
  switch(color){
    case ButtonColors.primary:
      return theme.colors.primary.main;
    case ButtonColors.danger:
      return theme.colors.danger.main;
    default: 
      return '#e0e0e0';
  }
}
const getColorText = ({theme, color})=>{
  switch(color){
    case ButtonColors.primary:
      return theme.colors.primary.text;
    case ButtonColors.danger:
      return theme.colors.danger.text;
    default: 
      return '#212121';
  }
};

const getDarkColor = ({theme, color})=>{
  switch(color){
    case ButtonColors.primary:
      return theme.colors.primary.dark;
    case ButtonColors.danger:
      return theme.colors.danger.dark;
    default: 
      return '#5a6268';
  }
}

const getOutlinedColor = (props) => {
  if(props.color===ButtonColors.default){
    return '#212121';
  }
  return getMainColor(props);
}

const getLinkText = (props) => {
  if(props.color===ButtonColors.default){
    return '#757575';
  }
  return getMainColor(props);
}

//não preciso pegar as props nesse caso, pois já esta vindo por padrão
const Button = styled.button`
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer; 
  background-color: ${getMainColor};
  border: 2px solid ${getMainColor};
  color: ${getColorText};
  display: inline-block;
  text-decoration: none;
  margin-right: 18px;
  &:disabled{
    opacity: 0.6;
    cursor: not-allowed;
  }
  &:hover:not(:disabled){ /**so aplica o hover se tiver habilitado */
    background-color: ${getDarkColor};
    border-color: 1px solid ${getDarkColor};
  }
`;
//herdando button e seus estilos
const ButtonOutlined = styled(Button)`
  background-color: transparent;
  color: ${getOutlinedColor};
  &:hover:not(:disabled){
    background-color: transparent;
    color: ${getDarkColor};
  }
`;
const ButtonLink = styled(Button)`
  background-color: transparent;
  border-color: transparent;
  color: ${getLinkText};
  padding-left: 0;
  padding-right: 0;

  &:hover:not(:disabled){
    background-color: transparent;
    border-color: transparent;
    color: ${getDarkColor};
  }
`;
//bg esta pegando a propriedade do themer
// const Button = ({type, children}) => (
//   <button type={type}>
//     {children}
//   </button>
// )

//necessario, pois o propTypes não tem compatibilidade com o styled-compoments
const ButtonWrapper = (props) => {
    switch (props.variant) {
      case ButtonsVariants.outlined:
        return <ButtonOutlined {...props}/>;
      case ButtonsVariants.link:
        return <ButtonLink {...props}/>
      default:
        return <Button {...props} />};
}

//aqui define as propriedades padrão do component
ButtonWrapper.defaultProps = {
  type: 'button',
  children: undefined,
  color: 'default',
  variante: 'default',
}
//aqui definimos o tipo e propriedades aceitas pelo component
ButtonWrapper.propTypes = {
  type: PropTypes.string,
  children: PropTypes.node,
  color: PropTypes.oneOf(Object.values(ButtonColors)),
  variant: PropTypes.oneOf(Object.values(ButtonsVariants)),
};

export default ButtonWrapper;
