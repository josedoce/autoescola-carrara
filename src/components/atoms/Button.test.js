import React from 'react';
import { render, fireEvent } from '../../test-utils';
import Button, { ButtonColors, ButtonsVariants } from './Button';

test('renders a text',() =>{
  const { getByText } = render(<Button>Click here</Button>);
  expect(getByText('Click here')).toBeInTheDocument();
});

//teste de evento de click
test('trigger event on click', () => {
  const handleClick = jest.fn();
  
  const {getByRole} = render(<Button onClick={handleClick}/>);

  fireEvent.click(getByRole('button'));

  expect(handleClick).toBeCalled();
});

//teste de caso com as cores
test.each(Object.values(ButtonColors).map(item=> [item]))('renders with color %s',(color) => { //%s de string
  const { asFragment } = render(<Button color={color}/>);
  expect(asFragment()).toMatchSnapshot();
});

//teste de caso com os variants
test.each(Object.values(ButtonsVariants).map(item=> [item]))('renders with variant %s',(variant) => { //%s de string
  const { asFragment } = render(<Button variant={variant}/>);
  expect(asFragment()).toMatchSnapshot();
});

//teste de caso de retorno padrao 1
test('Renders with color primary and variant outlined',() => {
  const { asFragment } = render(<Button variant='outlined' color='primary'/>);
  expect(asFragment()).toMatchSnapshot();
});

//teste de caso de retorno padrao 2
test('Renders with color primary and variant link',() => {
  const { asFragment } = render(<Button variant='link' color='primary'/>);
  expect(asFragment()).toMatchSnapshot();
});

