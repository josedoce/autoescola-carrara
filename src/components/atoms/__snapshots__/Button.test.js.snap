// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with color primary and variant link 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ sc-dlfnbm kzYLne iStTsS"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`Renders with color primary and variant outlined 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ sc-gsTCUz kzYLne fqKVRQ"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with color danger 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ ekmbQa"
    color="danger"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with color default 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ drdGNm"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with color primary 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ kzYLne"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with variant default 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ drdGNm"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with variant link 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ sc-dlfnbm drdGNm gzhhnD"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with variant outlined 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ sc-gsTCUz drdGNm lgAtkk"
    color="default"
    type="button"
  />
</DocumentFragment>
`;
